modules/features/discover/src/main/java/au/com/shiftyjelly/pocketcasts/discover/view/CategoriesListRowRedesignAdapter.kt

```diff
@@ -0,0 +1,151 @@
+package au.com.shiftyjelly.pocketcasts.discover.view
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.graphics.Color
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.core.content.ContextCompat.getDrawable
+import androidx.core.content.ContextCompat.getString
+import androidx.core.view.isVisible
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import androidx.recyclerview.widget.RecyclerView
+import au.com.shiftyjelly.pocketcasts.discover.R
+import au.com.shiftyjelly.pocketcasts.discover.databinding.CategoryPillBinding
+import au.com.shiftyjelly.pocketcasts.servers.model.DiscoverCategory
+import au.com.shiftyjelly.pocketcasts.servers.model.NetworkLoadableList
+
+private val CATEGORY_REDESIGN_DIFF = object : DiffUtil.ItemCallback<CategoryPill>() {
+    override fun areItemsTheSame(oldItem: CategoryPill, newItem: CategoryPill): Boolean {
+        return oldItem == newItem
+    }
+
+    override fun areContentsTheSame(oldItem: CategoryPill, newItem: CategoryPill): Boolean {
+        return oldItem.hashCode() == newItem.hashCode()
+    }
+}
+
+@SuppressLint("NotifyDataSetChanged")
```

Comment:
I have a general question about this class. The main benefit of using `ListAdapter` is that it handles diffing and computes output automatically. Using `notifyDataSetChanged()` on it is counterproductive. The way `ListAdapter` works best is that whenever an event happens it should pass it to the higher layers until the change reaches a view model. There, view model does necessary computations and updates items to be displayed by the adapter using `submitList()`.

This design where we keep internally `currentCategories` as a mutable list inside `submitList()` doesn't utilize this and makes it hard to test.

Reply:
I agree and I was not proud to use the `notifyDataSetChanged `. I tried this new implementation and I think it's better: [516c2ce](https://github.com/Automattic/pocket-casts-android/pull/1945/commits/516c2ced7cdb1676ad5ddcb32967e6f08eacaa49)

Thanks